pkg/api/testapi/testapi.go

```diff
@@ -24,63 +24,144 @@ import (
+			panic(err)
+		}
+		return interfaces.Codec
+	}
+	panic(fmt.Errorf("cannot test group %s", g.Group))
 }
 
 // Converter returns the api.Scheme for the API version to test against, as set by the
-// KUBE_API_VERSION env var.
-func Converter() runtime.ObjectConvertor {
-	interfaces, err := latest.InterfacesFor(Version())
-	if err != nil {
-		panic(err)
+// KUBE_TEST_API env var.
+func (g TestGroup) Converter() runtime.ObjectConvertor {
+	if g.Group == "" {
+		interfaces, err := latest.InterfacesFor(g.GroupVersionUnderTest)
+		if err != nil {
+			panic(err)
+		}
+		return interfaces.ObjectConvertor
+	}
+	if g.Group == "experimental" {
+		interfaces, err := explatest.InterfacesFor(g.GroupVersionUnderTest)
+		if err != nil {
+			panic(err)
+		}
+		return interfaces.ObjectConvertor
 	}
-	return interfaces.ObjectConvertor
+	panic(fmt.Errorf("cannot test group %s", g.Group))
+
 }
 
 // MetadataAccessor returns the MetadataAccessor for the API version to test against,
-// as set by the KUBE_API_VERSION env var.
-func MetadataAccessor() meta.MetadataAccessor {
-	interfaces, err := latest.InterfacesFor(Version())
-	if err != nil {
-		panic(err)
+// as set by the KUBE_TEST_API env var.
+func (g TestGroup) MetadataAccessor() meta.MetadataAccessor {
+	if g.Group == "" {
+		interfaces, err := latest.InterfacesFor(g.GroupVersionUnderTest)
+		if err != nil {
+			panic(err)
+		}
+		return interfaces.MetadataAccessor
+	}
+	if g.Group == "experimental" {
+		interfaces, err := explatest.InterfacesFor(g.GroupVersionUnderTest)
+		if err != nil {
+			panic(err)
+		}
+		return interfaces.MetadataAccessor
 	}
-	return interfaces.MetadataAccessor
+	panic(fmt.Errorf("cannot test group %s", g.Group))
 }
 
 // SelfLink returns a self link that will appear to be for the version Version().
 // 'resource' should be the resource path, e.g. "pods" for the Pod type. 'name' should be
 // empty for lists.
-func SelfLink(resource, name string) string {
+func (g TestGroup) SelfLink(resource, name string) string {
 	if name == "" {
-		return fmt.Sprintf("/api/%s/%s", Version(), resource)
+		return fmt.Sprintf("/api/%s/%s", g.Version(), resource)
 	}
-	return fmt.Sprintf("/api/%s/%s/%s", Version(), resource, name)
+	return fmt.Sprintf("/api/%s/%s/%s", g.Version(), resource, name)
 }
 
 // Returns the appropriate path for the given prefix (watch, proxy, redirect, etc), resource, namespace and name.
 // For ex, this is of the form:
 // /api/v1/watch/namespaces/foo/pods/pod0 for v1.
-func ResourcePathWithPrefix(prefix, resource, namespace, name string) string {
-	path := "/api/" + Version()
+func (g TestGroup) ResourcePathWithPrefix(prefix, resource, namespace, name string) string {
+	var path string
+	if len(g.Group) == 0 {
+		path = "/api/" + g.Version()
+	} else {
+		// TODO: switch back once we have proper multiple group support
+		// path = "/apis/" + g.Group + "/" + Version(group...)
+		path = "/" + g.Group + "/" + g.Version()
+	}
+
 	if prefix != "" {
 		path = path + "/" + prefix
```

Comment:
Ah, I see. I guess it's really a custom verb then.

On Fri, Sep 4, 2015 at 10:17 AM, Chao Xu <notifications@github.com> wrote:

> In pkg/api/testapi/testapi.go
> <https://github.com/kubernetes/kubernetes/pull/13581#discussion_r38773517>
> :
>
> >  }
> >
> >  // Returns the appropriate path for the given prefix (watch, proxy, redirect, etc), resource, namespace and name.
> >  // For ex, this is of the form:
> >  // /api/v1/watch/namespaces/foo/pods/pod0 for v1.
> > -func ResourcePathWithPrefix(prefix, resource, namespace, name string) string {
> > -	path := "/api/" + Version()
> > +func (g TestGroup) ResourcePathWithPrefix(prefix, resource, namespace, name string) string {
> > +	var path string
> > +	if len(g.Group) == 0 {
> > +		path = "/api/" + g.Version()
> > +	} else {
> > +		// TODO: switch back once we have proper multiple group support
> > +		// path = "/apis/" + g.Group + "/" + Version(group...)
> > +		path = "/" + g.Group + "/" + g.Version()
> > +	}
> > +
> >  	if prefix != "" {
> >  		path = path + "/" + prefix
>
> The position seems correct. It's not the same prefix we discussed. Its
> name is confusing.
>
> a usage:
>
> https://github.com/kubernetes/kubernetes/blob/ecc6ce773b81c7f89c97e734ce700ebd9b8f533d/pkg/client/unversioned/namespaces_test.go#L173
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13581/files#r38773517>.
>